### PR TITLE
[DSCAlarm] Various Fixes and Enhancements

### PR DESCRIPTION
**Pull request by RSStephens**
* Fixed 'DSC Alarm bridge handler not available.  Cannot handle command
without bridge.' error upon binding startup
* The binding was not sending all received messages to the channel
panel_message.  This is fixed.
* Fixed DSC Alarm command 060 not working.
* Fixed some errors found by the Static Code Analysis Tool

Enhancements
* Added send_command channel to allow for sending DSC commands.
* Ability to select the protocol type when using the TCPServer bridge.
* Modified and cleaned up README.md.

Signed-off by: Russell Stephens rsstephens@gmail.com (github: RSStephens)

### CODE REVIEW

File: addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml [hunk 1]
```diff
@@ -14,6 +14,10 @@
 				<label>Reset Envisalink</label>
 				<description>Resets the Envisalink</description>
 			</channel>
+            <channel id="send_command" typeId="command">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml]: there seems to be something screwed up regarding the formatting here...

File: addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml [hunk 2]
```diff
@@ -45,4 +49,10 @@
 		<description>Reset Switch</description>
 	</channel-type>
 
+    <channel-type id="command">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml]: You must define channel types only once within a binding, i.e. there must not be a channel type definition per bridge xml file.

File: addons/binding/org.openhab.binding.dscalarm/README.md [hunk 3]
```diff
@@ -27,7 +25,9 @@ There are essentially no overall binding configuration settings that need to be
...
+The DSC Alarm binding incorporates several discovery modes in order to find DSC Alarm systems.  First, there is the Envisalink bridge discovery mode which performs a network query for any Envisalink adapters and adds them to the discovery inbox.  Second, there is The IT-100 bridge discovery mode which will search serial ports for any IT-100 adapters and add them to the discovery inbox.  The bridge discovery modes are started manually through PaperUI.  Third, after a bridge is discovered and available to OpenHAB, the binding will attempt to discover DSC Alarm things and add them to the discovery inbox.  The TCP Server bridge does not implement bridge discovery but will utilize thing discovery once it is online.
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.dscalarm/README.md]: typo: should be openHAB, not OpenHAB.

Comment (RSStephens) [inline addons/binding/org.openhab.binding.dscalarm/README.md]: I will look at these ASAP!  Thanks Kai!

File: addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/panel.xml [hunk 4]
```diff
@@ -1,15 +1,13 @@
...
 	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
 	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
 
 	<thing-type id="panel">
-
 		<supported-bridge-type-refs>
 			<bridge-type-ref id="envisalink" />
 			<bridge-type-ref id="it100" />
-            <bridge-type-ref id="tcpserver" />
+			<bridge-type-ref id="tcpserver" />
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/panel.xml]: 👍

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseBridgeHandler.java [hunk 5]
```diff
@@ -540,21 +541,87 @@ public synchronized void handleIncomingMessage(String incomingMessage) {
...
+
+        try {
+            DSCAlarmCode dscAlarmCode = DSCAlarmCode.getDSCAlarmCodeValue(command);
+
+            if (dscAlarmProtocol.equals(DSCAlarmProtocol.IT100_API) && dscAlarmCode.equals(DSCAlarmCode.KeySequence)) {
+                return sendKeySequence(data);
+            } else {
+                return sendCommand(dscAlarmCode, data);
+            }
+        } catch (Exception e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseBridgeHandler.java]: Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/EnvisalinkBridgeHandler.java [hunk 6]
```diff
@@ -101,8 +101,8 @@ public void openConnection() {
             logger.debug("openConnection(): Connecting to Envisalink ");
 
             tcpSocket = new Socket();
-            SocketAddress TPIsocketAddress = new InetSocketAddress(ipAddress, tcpPort);
-            tcpSocket.connect(TPIsocketAddress, connectionTimeout);
+            SocketAddress tpiSocketAddress = new InetSocketAddress(ipAddress, tcpPort);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/EnvisalinkBridgeHandler.java]: 👍

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/KeypadThingHandler.java [hunk 7]
```diff
@@ -98,7 +96,12 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
      */
     @Override
     public void handleCommand(ChannelUID channelUID, Command command) {
-        // No Commands to Handle
+
+        logger.debug("handleCommand(): Command Received - {} {}.", channelUID, command);
+
+        if (command instanceof RefreshType) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/KeypadThingHandler.java]: What is the added value of this

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java [hunk 8]
```diff
@@ -180,76 +182,70 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
-                        case 10:
-                            dscAlarmBridgeHandler.sendCommand(DSCAlarmCode.SetTimeDate);
-                            break;
-                        case 200:
-                            dscAlarmBridgeHandler.sendCommand(DSCAlarmCode.CodeSend, getUserCode());
-                            break;
-                        default:
-                            break;
-                    }
+        if (dscAlarmBridgeHandler != null) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java]: Better make this a single if with an &&

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java [hunk 9]
```diff
@@ -180,76 +182,70 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
+
+            if (dscAlarmBridgeHandler.isConnected()) {
+                int cmd;
+
+                switch (channelUID.getId()) {
+                    case PANEL_COMMAND:
+                        cmd = Integer.parseInt(command.toString());
+                        switch (cmd) {
+                            case 0:
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java]: To me this values look quite random it's a common best practice to create named constants in such cases.

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java [hunk 10]
```diff
@@ -180,76 +182,70 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
-                    }
+        if (dscAlarmBridgeHandler != null) {
+
+            if (dscAlarmBridgeHandler.isConnected()) {
+                int cmd;
+
+                switch (channelUID.getId()) {
+                    case PANEL_COMMAND:
+                        cmd = Integer.parseInt(command.toString());
+                        switch (cmd) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PanelThingHandler.java]: Better not nest switches, maybe just move the inner switch to its own method.

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PartitionThingHandler.java [hunk 11]
```diff
@@ -97,37 +98,42 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
-                    } else if (command.toString().equals("5")) {
-                        dscAlarmBridgeHandler.sendCommand(DSCAlarmCode.PartitionArmControlWithUserCode,
-                                String.valueOf(partitionNumber));
-                    }
-                    break;
-                default:
-                    break;
+        if (dscAlarmBridgeHandler != null) {
+
+            if (dscAlarmBridgeHandler.isConnected()) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/PartitionThingHandler.java]: Better join with other if

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java [hunk 12]
```diff
@@ -100,26 +101,31 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
-                        String data = String.valueOf(getPartitionNumber()) + "*1"
-                                + String.format("%02d", getZoneNumber()) + "#";
-                        dscAlarmBridgeHandler.sendCommand(DSCAlarmCode.KeySequence, data);
-                    }
-                    break;
-                default:
-                    break;
+        if (dscAlarmBridgeHandler != null) {
+
+            if (dscAlarmBridgeHandler.isConnected()) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java]: Join

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java [hunk 13]
```diff
@@ -100,26 +101,31 @@ public void updateChannel(ChannelUID channelUID, int state, String description)
...
+                switch (channelUID.getId()) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java]: You can maybe join the `channelUID.getId() == ZONE_BYPASS_MODE` with the outer if no reason to use an if with a single case.

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/DSCAlarmMessage.java [hunk 14]
```diff
@@ -22,7 +22,7 @@
 
     private final Logger logger = LoggerFactory.getLogger(DSCAlarmMessage.class);
 
-    private static final EnumMap<DSCAlarmCode, MessageParameters> dscAlarmMessageParameters = new EnumMap<>(
+    private static final EnumMap<DSCAlarmCode, MessageParameters> DSCALARM_MESSAGE_PARAMETERS = new EnumMap<>(
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/DSCAlarmMessage.java]: 👍

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java (hunk 13 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java]: I mean the == not literal I assume it should use equals

File: addons/binding/org.openhab.binding.rfxcom.test/.classpath [hunk 15]
```diff
@@ -7,5 +7,6 @@
 	<classpathentry combineaccessrules="false" kind="src" path="/org.eclipse.smarthome.core"/>
 	<classpathentry combineaccessrules="false" kind="src" path="/org.eclipse.smarthome.core.thing"/>
 	<classpathentry combineaccessrules="false" kind="src" path="/org.eclipse.smarthome.core.thing.xml.test"/>
+	<classpathentry exported="true" kind="con" path="GROOVY_DSL_SUPPORT"/>
```
Comment (RSStephens) [inline addons/binding/org.openhab.binding.rfxcom.test/.classpath]: No this is definitely not apart of this PR.  Not sure how that happened.

File: addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseBridgeHandler.java (hunk 5 above)
Comment (RSStephens) [inline addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseBridgeHandler.java]: I can probably remove the try-catch statements.

**Comment by RSStephens** [comment]
Hello @martinvw,

I will look into making the changes as soon as possible.  Thank you!

**Comment by martinvw** [comment]
Thanks, now we only have one crappy detail all your commits lack a signed-off-by and you merged master in a few times, it's gonna be tricky to recover from this.

If you are not comfortable doing that, the easiest thing to do is apply all changes to a fresh copy of the master and force pushing it here.

- start with creating a new branch based on your existing (this) branch and push it GitHub to check whether it's okay, that is your backup :-)
- locally remove the branch with the name 'DSCAlarm' but not the backup
- switch to master and update the master to the latest version
- recreate the 'DSCAlarm' branch, but do not yet push it
- check for a proper tool to apply diffs or patches for your platform a lot of git tools can do it and I suppose your also do it.
- download a diff or patch file from GitHub for the changes in your branch you can download them by adding .diff or .patch to the pr URL https://github.com/openhab/openhab2-addons/pull/2057
- apply it and force push your branch to GitHub validate it before removing the backup.

Questions just let me know

**Comment by RSStephens** [comment]
Hello Martin,

Could I close this PR and submit a clean PR after straightening things out?

**Comment by martinvw** [comment]
Okay, with me. Please do post a link to this pr so we know the history

**Comment by RSStephens** [comment]
Closing this PR to resubmit new one.

**Pull request closed without merging**